src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class TimeSpanConverter : JsonConverter<TimeSpan>
+    {
+        private const int MinimumTimeSpanFormatLength = 7; // h:mm:ss
+        private const int MaximumTimeSpanFormatLength = 26; // -dddddddd.hh:mm:ss.fffffff
+
+        public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw ThrowHelper.GetInvalidOperationException_ExpectedString(reader.TokenType);
+            }
+
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+
+            if (reader.HasValueSequence)
+            {
+                ReadOnlySequence<byte> valueSequence = reader.ValueSequence;
+                long sequenceLength = valueSequence.Length;
+
+                if (!JsonHelpers.IsInRangeInclusive(sequenceLength, MinimumTimeSpanFormatLength, MaximumTimeSpanFormatLength))
+                {
+                    throw ThrowHelper.GetFormatException();
+                }
+
+                Span<byte> stackSpan = stackalloc byte[(int)sequenceLength];
+
+                valueSequence.CopyTo(stackSpan);
+                source = stackSpan;
+            }
+            else
+            {
+                source = reader.ValueSpan;
+
+                if (!JsonHelpers.IsInRangeInclusive(source.Length, MinimumTimeSpanFormatLength, MaximumTimeSpanFormatLength))
+                {
+                    throw ThrowHelper.GetFormatException();
+                }
+            }
+
+            bool result = Utf8Parser.TryParse(source, out TimeSpan tmpValue, out int bytesConsumed, 'c');
+
+            // Note: Utf8Parser.TryParse will return true for invalid input so
+            // long as it starts with an integer. Example: "2021-06-18" or
+            // "1$$$$$$$$$$". We need to check bytesConsumed to know if the
+            // entire source was actually valid.
+
+            if (result && source.Length == bytesConsumed)
+            {
+                return tmpValue;
+            }
+
+            result = Utf8Parser.TryParse(source, out tmpValue, out bytesConsumed, 'g');
```

Comment:
Does this also support [The General Long ("G") Format Specifier](https://docs.microsoft.com/dotnet/standard/base-types/standard-timespan-format-strings#the-general-long-g-format-specifier)?

Reply:
You can pass 'G' to `Utf8Parser.TryParse` to get a specific ["BigG" parser](https://github.com/dotnet/runtime/blob/fc4a42746d9cf65ad3db4074207069ab9c168bc9/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.TimeSpan.BigG.cs#L8) but I don't _think_ it is necessary here. From a reading perspective, "G" is basically a "c" with everything specified. So I thinik the initial "c" call to `Utf8Parser.TryParse` will also pick up the "G"s. Where I am not sure is how the culture plays into the separator. I don't think `Utf8Parser` handles that whatsoever and [forces '.' for separators](https://github.com/dotnet/runtime/blob/de5dd0d370cda957cf36b687a066e34411bc4f7c/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.TimeSpanSplitter.cs#L179).

Reply:
https://github.com/dotnet/runtime/pull/54186#pullrequestreview-694153448.